torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -91,6 +106,376 @@ def _reshard_grads(
+    with torch.autograd.profiler.record_function(
+        "FullyShardedDataParallel._post_backward_hook"
+    ):
+        _assert_in_training_states(state, [TrainingState.FORWARD_BACKWARD])
+        state.training_state = TrainingState.FORWARD_BACKWARD
+        p_assert(
+            handle._training_state == HandleTrainingState.BACKWARD_PRE,
+            f"Expects `BACKWARD_PRE` state but got {handle._training_state}",
+        )
+        handle._training_state = HandleTrainingState.BACKWARD_POST
+
+        if param.grad is None:
+            return
+        if param.grad.requires_grad:
+            raise RuntimeError("FSDP does not support gradients of gradients")
+
+        free_unsharded_flat_param = _should_free_in_backward(state, handle)
+        _reshard(state, [handle], [free_unsharded_flat_param])
+
+        # TODO: Post-backward prefetching does not support the multiple handles
+        # per module case since the post-backward hook runs per handle, not per
+        # group of handles.
+        handles_key = (handle,)
+        _prefetch_handles(state, handles_key)
+
+        if not state._sync_gradients:
+            return
+
+        # Wait for all ops in the current stream (e.g. gradient
+        # computation) to finish before reduce-scattering the gradient
+        state._streams["post_backward"].wait_stream(torch.cuda.current_stream())
+
+        with torch.cuda.stream(state._streams["post_backward"]):
+            unsharded_grad_data = param.grad.data
+            if state._exec_order_data.is_first_iter:  # only check once
+                _check_comm_hook(
+                    state._communication_hook, state._communication_hook_state
+                )
+            if handle._uses_reduce_mixed_precision and not _low_precision_hook_enabled(
+                state
+            ):
+                # TODO: Use the low precision communication hook directly
+                param.grad.data = param.grad.to(state.mixed_precision.reduce_dtype)
+
+            if handle.uses_sharded_strategy:
+                # We clear `.grad` to permit multiple backwards. This avoids a
+                # race where the second backward pass computation precedes
+                # ahead of the first backward pass reduction, which is possible
+                # since the reduction is issued in a separate stream and is
+                # async and would result in reducing the wrong gradient.
+                unsharded_grad = param.grad.data
+                param.grad = None
+                p_assert(
+                    len(unsharded_grad.size()) == 1,
+                    f"Expects gradient to be flattened but got {unsharded_grad.size()}",
+                )
+                chunks = list(unsharded_grad.chunk(state.world_size))
+                numel_to_pad = (
+                    state.world_size * chunks[0].numel() - unsharded_grad.numel()
+                )
+                padded_unsharded_grad = F.pad(unsharded_grad, [0, numel_to_pad])
```

Comment:
will this create a new and larger padded tensor for unsharded grad? If yes, it seems more reasonable to keep the padded one and `param.grad` points to its view?